### ISSUE TITLE
Possible Firefox compatibility issues for local server

### ISSUE THREAD

**Issue by igreen1** [dondi/GRNsight#891]
When attempting to access server on port 5001 from port 5000, Firefox gives the following error (Webclient is on port 5000 and server is on port 5001)

When trying to upload files:
`Cross-Origin Request Blocked: The Same Origin Policy disallows reading the remote resource at http://localhost:5000/upload-graphml. (Reason: CORS request did not succeed)`

When trying to load demo files:
`Cross-Origin Request Blocked: The Same Origin Policy disallows reading the remote resource at http://localhost:5000/demo/unweighted. (Reason: CORS request did not succeed).`

Any attempt to access server resources throw this error. I am able to load demo files on Chrome, which implies to me the error occurs because of Firefox. This error appeared recently. In my testing, I tried to disable all extensions, disable some extensions, enter private browsing and any combination therein I could think of.

The Docs
-------------------------------------------------------

This is the exact error, in the Mozilla docs. To clarify, again , this error occurs with no extensions (plugins in Mozilla lingo)
https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS/Errors/CORSDidNotSucceed

"As of Firefox 68, http pages are not permitted to access http://localhost"

The above is likely involved in the error, but it doesn't explain the recent change in functionality. It worked on Firefox 80.0 without issues, and sometimes worked on Firefox 81.0, but as of 82.0 it is entirely broken.

Some notes below
------------------------------------------------------

As an initial investigation, Firefox has introduced many many new privacy restrictions to block tracking cookies. A recent (ish) addition was Enhanced Tracking Protection 2.0 which 'blocks redirect trackers by default' which could lead to these issues.
(https://blog.mozilla.org/blog/2020/08/04/latest-firefox-rolls-out-enhanced-tracking-protection-2-0-blocking-redirect-trackers-by-default/)

Firefox Version 82.0, which I am currently using, was released October 20. Firefox version 81.0 had similar issues but only on Windows. So, the issue was probably released somewhere after Firefox version 80. 

The best resource I found (from Mozilla themselves)
https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS/Errors

**Comment by dondi** [comment]
In my tests with Firefox 82.0.2 on macOS, the two routes mentioned in the description, _upload-graphml_ and _demo/unweighted_, both loaded successfully without CORS errors. I was using NodeJS 14.13.1. (the GraphML load didn’t complete anyway due to the `network.meta` bug, but the network connection itself did succeed and didn’t encounter a CORS error)

I might have to do this on a screenshare with @igreen1—until I see the errors firsthand, it’s difficult to triage this further.

**Comment by dondi** [comment]
CORS errors could not be replicated on Ubuntu either. I might need to screenshare this to make sure that I’m doing the exact steps. The same behavior was noted: successful network connections in both cases, with the expected `network.meta` bug happening after the fact. No signs of CORS failures.

**Comment by igreen1** [comment]
I have been able replicate this issue. Some notes

1. EFF's privacy badger breaks the site. It blocks a variety of trackers that are built into the app, and essential. Specifically, it blocks attempts to access cdnjs.cloudflare.com, which will always break the app. It also blocks google-analytics, which as described below will sometimes break the app. I think this is my issue. I had merely disabled EFFPB for 'this session' instead of permanently, so reloads kept messing up.

2. Firefox default blocks google-analytics, which /sometimes/ breaks the app. I can't quite tell why, but when google-analytics is off, the app sometimes won't load. I'd guess there is some async function not working. It's rare and a simple refresh will fix the issue. Whitelisting google-analytics also fixes the issue.

3. Firefox in strict privacy mode will /sometimes/ break the app even with google-analytics whitelisted. I'd guess its the same issue as (1), where cdnjs.cloudflare.com is blocked. Frequent reloading contributes to the issue. Its hard to tell exaclty though, as its difficult to reliably replicate.

4. UBlock origin (an adblock software) will ALSO block google-analytics, which again sometimes breaks the app. UBlock origin in strict mode also sometimes blocks the cdnsj.cloudfare.com access, which always break the app.

5. Placeholder. I'm writing this comment before testing on Windows so I'll put updates on windows here.

In summary, the app breaking appears to be a conflict with our trackers (or the perception of a tracker) and Firefox / some extensions attempting to block all trackers. This causes wildly unpredictable behaviour that is hard to pinpoint. 

I'd say we should just add a simple note in the dev-wiki to say "disable privacy software and extensions."

**Comment by dondi** [comment]
Thanks @igreen1—those are sensible findings (and they also explain why I haven't been able to replicate these issues). They do seem to be on the aggressive side (not all CORS requests are tracking requests) but it’s possible that they have both whitelists and blacklists and, although we wouldn't be on a blacklist, we might also not be on a whitelist so we get blocked.

The note about extensions sounds reasonable and I think some version of it will suffice in terms of addressing this issue. We can discuss this at the meeting.

**Comment by dondi** [comment]
Upon some inspection of the code and discussion of @igreen1's notes, it looks like GRNsight’s GA counter (which tracks how many workbooks have been uploaded) occasionally encounters an error due to the more aggressive blocking rules of newer browsers + plugins which disrupts the normal operation of GRNsight. Ideally, the only consequence of such an error is that the GA counter doesn’t update but GRNsight still continues with functionality.

@igreen1 will take these leads to see if he can track down some replicable scenario where the blockage disrupts GRNsight and thus we might be able to write some mitigating code around that.

**Comment by dondi** [comment]
@igreen1 reports that this is no longer replicable and may be due to activating a new “Strict mode” (for privacy) in Firefox.
